File: backend/app/routers/google_slides.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session
from app.database import get_db
from app.config.google_slides import get_auth_url, set_auth_token, get_google_slides_client
from app.services.google_slides_parser import fetch_presentation, extract_slide_elements
from app.services.slides_to_html import convert_presentation_to_html, convert_slide_to_html
from app.services.gemini_reviewer import review_presentation, review_slide_html
from app.services.db.template_service import create_template
import re

router = APIRouter(prefix="/api/google-slides", tags=["google-slides"])


class PresentationIdRequest(BaseModel):
    presentation_id: str


class URLRequest(BaseModel):
    url: str


class ReviewSlideRequest(BaseModel):
    html: str
    slide_index: int


@router.get("/auth")
async def get_auth():
    """Google認証URLを取得"""
    try:
        auth_url = get_auth_url()
        return {
            "success": True,
            "data": {
                "authUrl": auth_url,
            },
        }
    except Exception as e:
        print(f"Error generating auth URL: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to generate auth URL: {str(e)}",
        )


@router.get("/callback")
async def callback(code: str):
    """OAuthコールバック"""
    try:
        if not code:
            raise HTTPException(status_code=400, detail="Authorization code is required")
        
        refresh_token = set_auth_token(code)
        
        return {
            "success": True,
            "message": "Authentication successful. Please save the refresh token to your .env file.",
            "refresh_token": refresh_token,
        }
    except Exception as e:
        print(f"Error in OAuth callback: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Authentication failed: {str(e)}",
        )


@router.post("/import")
async def import_presentation(
    request: PresentationIdRequest,
    db: Session = Depends(get_db),
):
    """Google SlidesプレゼンテーションをインポートしてHTMLテンプレートを生成"""
    try:
        # プレゼンテーションを取得
        presentation = fetch_presentation(request.presentation_id)
        
        if not presentation.get("slides"):
            raise HTTPException(status_code=400, detail="No slides found in presentation")
        
        # HTMLに変換
        html = convert_presentation_to_html(presentation)
        
        # 各スライドのHTMLを抽出
        slides_html: list[str] = []
        for i, slide in enumerate(presentation.get("slides", [])):
            elements = extract_slide_elements(slide)
            slide_html = convert_slide_to_html(elements, i)
            slides_html.append(slide_html)
        
        # Geminiでレビュー
        review = await review_presentation(slides_html)
        
        # スロットを生成
        slots = []
        for slide_index, slide_review in enumerate(review.slide_reviews):
            for slot in slide_review.suggested_slots:
                slots.append({
                    "slide_index": slide_index,
                    "name": slot.name,
                    "type": slot.type,
                    "placeholder": slot.placeholder,
                    "required": slot.required,
                })
        
        # テンプレートをDBに保存
        saved_template = create_template(
            db,
            name=review.title,
            description=review.description,
            thumbnail=presentation.get("title", "") or "",
            slide_count=len(presentation.get("slides", [])),
            html=html,
            slots=slots,
        )
        
        return {
            "success": True,
            "data": {
                "template": saved_template.dict(),
                "html": html,
                "review": review.dict(),
            },
        }
    except Exception as e:
        print(f"Error importing Google Slides: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to import Google Slides: {str(e)}",
        )


@router.post("/parse-url")
async def parse_url(
    request: URLRequest,
    db: Session = Depends(get_db),
):
    """Google Slides URLからプレゼンテーションIDを抽出してインポート"""
    try:
        # URLからプレゼンテーションIDを抽出
        match = re.search(r"/presentation/d/([a-zA-Z0-9-_]+)", request.url)
        if not match or not match.group(1):
            raise HTTPException(status_code=400, detail="Invalid Google Slides URL")
        
        presentation_id = match.group(1)
        
        # インポート処理を実行
        presentation = fetch_presentation(presentation_id)
        html = convert_presentation_to_html(presentation)
        
        slides_html: list[str] = []
        for i, slide in enumerate(presentation.get("slides", [])):
            elements = extract_slide_elements(slide)
            slide_html = convert_slide_to_html(elements, i)
            slides_html.append(slide_html)
        
        review = await review_presentation(slides_html)
        
        slots = []
        for slide_index, slide_review in enumerate(review.slide_reviews):
            for slot in slide_review.suggested_slots:
                slots.append({
                    "slide_index": slide_index,
                    "name": slot.name,
                    "type": slot.type,
                    "placeholder": slot.placeholder,
                    "required": slot.required,
                })
        
        # テンプレートをDBに保存
        saved_template = create_template(
            db,
            name=review.title,
            description=review.description,
            thumbnail=presentation.get("title", "") or "",
            slide_count=len(presentation.get("slides", [])),
            html=html,
            slots=slots,
        )
        
        return {
            "success": True,
            "data": {
                "presentationId": presentation_id,
                "template": saved_template.dict(),
                "html": html,
                "review": review.dict(),
            },
        }
    except Exception as e:
        print(f"Error parsing Google Slides URL: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to parse Google Slides URL: {str(e)}",
        )


@router.post("/review-slide")
async def review_slide(request: ReviewSlideRequest):
    """単一スライドのHTMLをレビュー"""
    try:
        review = await review_slide_html(request.html, request.slide_index)
        
        return {
            "success": True,
            "data": review.dict(),
        }
    except Exception as e:
        print(f"Error reviewing slide: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to review slide: {str(e)}",
        )

File: backend/src/routes/google-slides.ts
import express from "express"
import { z } from "zod"
import { getAuthUrl, setAuthToken } from "../config/google-slides.js"
import { fetchPresentation } from "../services/google-slides-parser.js"
import { convertPresentationToHTML, convertSlideToHTML } from "../services/slides-to-html.js"
import { extractSlideElements } from "../services/google-slides-parser.js"
import { reviewPresentation, reviewSlideHTML } from "../services/gemini-reviewer.js"

const router = express.Router()

/**
 * GET /api/google-slides/auth
 * Google認証URLを取得
 */
router.get("/auth", (req, res) => {
  try {
    const authUrl = getAuthUrl()
    res.json({
      success: true,
      data: {
        authUrl,
      },
    })
  } catch (error) {
    console.error("Error generating auth URL:", error)
    res.status(500).json({
      success: false,
      error: "Failed to generate auth URL",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

/**
 * GET /api/google-slides/callback
 * OAuthコールバック
 */
router.get("/callback", async (req, res) => {
  try {
    const { code } = req.query

    if (!code || typeof code !== "string") {
      return res.status(400).json({
        success: false,
        error: "Authorization code is required",
      })
    }

    await setAuthToken(code)

    res.json({
      success: true,
      message: "Authentication successful. Please save the refresh token to your .env file.",
    })
  } catch (error) {
    console.error("Error in OAuth callback:", error)
    res.status(500).json({
      success: false,
      error: "Authentication failed",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

/**
 * POST /api/google-slides/import
 * Google SlidesプレゼンテーションをインポートしてHTMLテンプレートを生成
 */
router.post("/import", async (req, res) => {
  try {
    const schema = z.object({
      presentationId: z.string(),
    })

    const { presentationId } = schema.parse(req.body)

    // プレゼンテーションを取得
    const presentation = await fetchPresentation(presentationId)

    if (!presentation.slides) {
      return res.status(400).json({
        success: false,
        error: "No slides found in presentation",
      })
    }

    // HTMLに変換
    const html = convertPresentationToHTML(presentation)

    // 各スライドのHTMLを抽出
    const slidesHTML: string[] = []
    for (let i = 0; i < (presentation.slides?.length || 0); i++) {
      const slide = presentation.slides[i]
      if (slide) {
        const elements = extractSlideElements(slide)
        const slideHTML = convertSlideToHTML(elements, i)
        slidesHTML.push(slideHTML)
      }
    }

    // Geminiでレビュー
    const review = await reviewPresentation(slidesHTML)

    // スロットを生成
    const slots = review.slideReviews.flatMap((slideReview, slideIndex) =>
      slideReview.suggestedSlots.map((slot) => ({
        ...slot,
        slideIndex,
      }))
    )

    // テンプレートをDBに保存
    const { createTemplate } = await import("../services/db/template.service.js")
    const savedTemplate = await createTemplate({
      name: review.title,
      description: review.description,
      thumbnail: presentation.thumbnail || "",
      slideCount: presentation.slides?.length || 0,
      html,
      slots: slots.map((slot) => ({
        slideIndex: slot.slideIndex,
        name: slot.name,
        type: slot.type,
        placeholder: slot.placeholder,
        required: slot.required,
      })),
    })

    res.json({
      success: true,
      data: {
        template: {
          id: savedTemplate.id,
          name: savedTemplate.name,
          description: savedTemplate.description,
          thumbnail: savedTemplate.thumbnail,
          slideCount: savedTemplate.slideCount,
          slots: savedTemplate.slots,
          createdAt: savedTemplate.createdAt,
        },
        html,
        review,
      },
    })
  } catch (error) {
    console.error("Error importing Google Slides:", error)
    res.status(500).json({
      success: false,
      error: "Failed to import Google Slides",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

/**
 * POST /api/google-slides/parse-url
 * Google Slides URLからプレゼンテーションIDを抽出してインポート
 */
router.post("/parse-url", async (req, res) => {
  try {
    const schema = z.object({
      url: z.string().url(),
    })

    const { url } = schema.parse(req.body)

    // URLからプレゼンテーションIDを抽出
    // https://docs.google.com/presentation/d/{PRESENTATION_ID}/edit
    const match = url.match(/\/presentation\/d\/([a-zA-Z0-9-_]+)/)
    if (!match || !match[1]) {
      return res.status(400).json({
        success: false,
        error: "Invalid Google Slides URL",
      })
    }

    const presentationId = match[1]

    // インポート処理を実行
    const importResponse = await fetchPresentation(presentationId)
    const html = convertPresentationToHTML(importResponse)

    const slidesHTML: string[] = []
    if (importResponse.slides) {
      for (let i = 0; i < importResponse.slides.length; i++) {
        const slide = importResponse.slides[i]
        if (slide) {
          const elements = extractSlideElements(slide)
          const slideHTML = convertSlideToHTML(elements, i)
          slidesHTML.push(slideHTML)
        }
      }
    }

    const review = await reviewPresentation(slidesHTML)

    const slots = review.slideReviews.flatMap((slideReview, slideIndex) =>
      slideReview.suggestedSlots.map((slot) => ({
        ...slot,
        slideIndex,
      }))
    )

    // テンプレートをDBに保存
    const { createTemplate } = await import("../services/db/template.service.js")
    const savedTemplate = await createTemplate({
      name: review.title,
      description: review.description,
      thumbnail: importResponse.thumbnail || "",
      slideCount: importResponse.slides?.length || 0,
      html,
      slots: slots.map((slot) => ({
        slideIndex: slot.slideIndex,
        name: slot.name,
        type: slot.type,
        placeholder: slot.placeholder,
        required: slot.required,
      })),
    })

    res.json({
      success: true,
      data: {
        presentationId,
        template: {
          id: savedTemplate.id,
          name: savedTemplate.name,
          description: savedTemplate.description,
          thumbnail: savedTemplate.thumbnail,
          slideCount: savedTemplate.slideCount,
          slots: savedTemplate.slots,
          createdAt: savedTemplate.createdAt,
        },
        html,
        review,
      },
    })
  } catch (error) {
    console.error("Error parsing Google Slides URL:", error)
    res.status(500).json({
      success: false,
      error: "Failed to parse Google Slides URL",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

/**
 * POST /api/google-slides/review-slide
 * 単一スライドのHTMLをレビュー
 */
router.post("/review-slide", async (req, res) => {
  try {
    const schema = z.object({
      html: z.string(),
      slideIndex: z.number().int().min(0),
    })

    const { html, slideIndex } = schema.parse(req.body)

    const review = await reviewSlideHTML(html, slideIndex)

    res.json({
      success: true,
      data: review,
    })
  } catch (error) {
    console.error("Error reviewing slide:", error)
    res.status(500).json({
      success: false,
      error: "Failed to review slide",
      message: error instanceof Error ? error.message : "Unknown error",
    })
  }
})

export default router

